File: Archiver/app_code/clsDBUpdate.cs
using System;
using System.Collections.Generic;
using global::System.IO;
using System.Windows.Forms;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using MODI;

namespace EcmArchiver
{
    public class clsDBUpdate
    {
        private List<FileInfo> UpdateINfo = new List<FileInfo>();
        private List<string> FilterList = new List<string>();
        private clsDatabaseARCH DB = new clsDatabaseARCH();
        private clsLogging LOG = new clsLogging();

        public bool CheckForDBUpdates()
        {
            My.MyProject.Forms.frmDBUpdates.Text = "Applying Database Updates";
            FilterList.Clear();
            FilterList.Add("*.sql");
            string FLDR = AppDomain.CurrentDomain.BaseDirectory;
            string UPDTFOLDER = FLDR + @"\DBUpdates";
            if (!Directory.Exists(UPDTFOLDER))
            {
                return true;
            }

            string strFileSize = "";
            var di = new DirectoryInfo(UPDTFOLDER);
            var aryFi = di.GetFiles("*.sql");
            int iline = 0;
            foreach (var fi in aryFi)
            {
                string FName = fi.Name;
                string FQN = fi.FullName;
                var LastWriteTime = fi.LastWriteTime;
                int I = DB.getDBUpdateExists(FQN);
                My.MyProject.Forms.frmDBUpdates.SB.Text = fi.Name;
                My.MyProject.Forms.frmDBUpdates.Refresh();
                Application.DoEvents();
                if (I > 0)
                {
                    var DateApplied = DB.getDBUpdateLastWriteDate(FQN);
                    int secs = SecDiff(LastWriteTime, DateApplied);
                    // If LastWriteTime > DateApplied Then
                    if (secs > 0)
                    {
                        string fn = Path.GetFileName(FQN);
                        LOG.WriteToArchiveLog("APPLYING DB UPDATE: " + fn);
                        My.MyProject.Forms.frmDBUpdates.txtFile.Text = Path.GetFileName(FQN);
                        ApplyUpdate(FQN);
                        LOG.WriteToArchiveLog("COMPLETED DB UPDATE: " + fn);
                        Console.WriteLine("APPLIED");
                    }
                }
                else
                {
                    ApplyUpdate(FQN);
                }
            }

            return default;
        }

        public int SecDiff(DateTime LastWriteTime, DateTime DateApplied)
        {
            int secs = 0;
            var span = LastWriteTime.Subtract(DateApplied);
            return span.Seconds;
        }

        public int ApplyUpdate(string FQN)
        {
            string S = "";
            int Ifailures = 0;
            var reader = My.MyProject.Computer.FileSystem.OpenTextFileReader(FQN);
            string sline = "";
            string tline = "";
            string MySql = "";
            string cs = DB.getRepoConnStr();
            bool B = false;
            int iline = 0;
            using (reader)
            {
                sline = reader.ReadLine();
                do
                {
                    iline += 1;
                    Application.DoEvents();
                    sline = sline.Trim();
                    tline = sline.ToUpper();
                    if (tline.Equals("GO"))
                    {
                        My.MyProject.Forms.frmDBUpdates.txtSql.Text = Path.GetFileName(MySql);
                        My.MyProject.Forms.frmDBUpdates.Refresh();
                        Application.DoEvents();
                        LOG.WriteToDBUpdatesLog("--************************************");
                        LOG.WriteToDBUpdatesLog("--Update File:" + FQN);
                        LOG.WriteToDBUpdatesLog(MySql);
                        B = DB.ExecuteSql(MySql, cs, false);
                        if (B)
                        {
                            B = DB.insertDBUpdate(FQN);
                            if (B)
                            {
                                LOG.WriteToArchiveLog("DB Update applied from " + FQN);
                                B = MarkFileAsApplied(FQN, cs);
                                if (B)
                                {
                                    string FN = Path.GetFileName(FQN);
                                    LOG.WriteToArchiveLog("Notice DB Update:" + FN + " marked as APPLIED.");
                                    DB.updateDBUpdateLastwrite(FQN);
                                }
                            }
                        }
                        else
                        {
                            LOG.WriteToArchiveLog("ERROR: DB Update FAILED from " + FQN + Constants.vbCrLf + MySql);
                        }

                        MySql = "";
                    }
                    else
                    {
                        MySql += sline + Environment.NewLine;
                    }

                    sline = reader.ReadLine();
                }
                while (!(sline is null));
            }

            if (MySql.Trim().Length > 0)
            {
                My.MyProject.Forms.frmDBUpdates.txtSql.Text = Path.GetFileName(MySql);
                LOG.WriteToDBUpdatesLog("--************************************");
                LOG.WriteToDBUpdatesLog("--Update File:" + FQN);
                LOG.WriteToDBUpdatesLog(MySql);
                B = DB.ExecuteSql(MySql, cs, false);
                if (B)
                {
                    B = DB.insertDBUpdate(FQN);
                    if (B)
                    {
                        LOG.WriteToArchiveLog("DB Update applied from " + FQN);
                        B = MarkFileAsApplied(FQN, cs);
                        DB.updateDBUpdateLastwrite(FQN);
                        if (B)
                        {
                            string FN = Path.GetFileName(FQN);
                            LOG.WriteToArchiveLog("Notice DB  Update: " + FN + " marked as APPLIED.");
                        }
                    }
                }
                else
                {
                    LOG.WriteToArchiveLog("ERROR: DB Update FAILED from " + FQN + Constants.vbCrLf + MySql);
                }
            }

            B = MarkFileAsApplied(FQN, cs);
            My.MyProject.Forms.frmDBUpdates.txtSql.Text = "";
            return Ifailures;
        }

        public bool MarkFileAsApplied(string FQN, string cs)
        {
            bool b = true;
            FQN = FQN.Replace("'", "''");
            string MySql = "update DBUpdate set UpdateApplied = 1 where FileName = '" + FQN + "'";
            b = DB.ExecuteSql(MySql, cs, false);
            return b;
        }

        public List<FileInfo> GetFiles(string Path, List<string> FilterList)
        {
            var d = new DirectoryInfo(Path);
            var files = new List<FileInfo>();
            foreach (string Filter in FilterList)
            {
                // the files are appended to the file array
                Application.DoEvents();
                files.AddRange(d.GetFiles(Filter));
            }

            files.Sort(Operators.ConditionalCompareObjectGreaterEqual(f, f.LastWritetime, false));
            return files;
        }
    }
}

File: Archiver/app_code/clsDBUpdate.vb
Imports System.IO
Imports System.Data.SqlClient

Public Class clsDBUpdate

    Dim UpdateINfo As New List(Of FileInfo)
    Dim FilterList As New List(Of String)

    Dim DB As New clsDatabaseARCH
    Dim LOG As New clsLogging

    Public Function CheckForDBUpdates() As Boolean

        frmDBUpdates.Text = "Applying Database Updates"

        FilterList.Clear()
        FilterList.Add("*.sql")

        Dim FLDR As String = System.AppDomain.CurrentDomain.BaseDirectory()
        Dim UPDTFOLDER As String = FLDR + "\DBUpdates"

        Dim strFileSize As String = ""
        Dim di As New IO.DirectoryInfo(UPDTFOLDER)
        Dim aryFi As IO.FileInfo() = di.GetFiles("*.sql")
        Dim fi As IO.FileInfo
        Dim iline As Integer = 0

        For Each fi In aryFi

            Dim FName As String = fi.Name
            Dim FQN As String = fi.FullName
            Dim LastWriteTime As DateTime = fi.LastWriteTime
            Dim I As Integer = DB.getDBUpdateExists(FQN)

            frmDBUpdates.SB.Text = fi.Name
            frmDBUpdates.Refresh()
            Application.DoEvents()

            If I > 0 Then
                Dim DateApplied As DateTime = DB.getDBUpdateLastWriteDate(FQN)
                Dim secs As Integer = SecDiff(LastWriteTime, DateApplied)
                'If LastWriteTime > DateApplied Then
                If secs > 0 Then
                    Dim fn As String = Path.GetFileName(FQN)
                    LOG.WriteToArchiveLog("APPLYING DB UPDATE: " + fn)
                    frmDBUpdates.txtFile.Text = Path.GetFileName(FQN)
                    ApplyUpdate(FQN)
                    LOG.WriteToArchiveLog("COMPLETED DB UPDATE: " + fn)
                    Console.WriteLine("APPLIED")
                End If
            Else
                ApplyUpdate(FQN)
            End If
        Next

    End Function

    Function SecDiff(LastWriteTime As DateTime, DateApplied As DateTime) As Integer

        Dim secs As Integer = 0
        Dim span As TimeSpan = LastWriteTime.Subtract(DateApplied)

        Return span.Seconds

    End Function

    Function ApplyUpdate(FQN As String) As Integer


        Dim S As String = ""
        Dim Ifailures As Integer = 0

        Dim reader As StreamReader = My.Computer.FileSystem.OpenTextFileReader(FQN)
        Dim sline As String = ""
        Dim tline As String = ""
        Dim MySql As String = ""
        Dim cs As String = DB.getRepoConnStr
        Dim B As Boolean = False
        Dim iline As Integer = 0
        Using reader
            sline = reader.ReadLine
            Do
                iline += 1
                Application.DoEvents()
                sline = sline.Trim()
                tline = sline.ToUpper
                If tline.Equals("GO") Then
                    frmDBUpdates.txtSql.Text = Path.GetFileName(MySql)
                    frmDBUpdates.Refresh()
                    Application.DoEvents()
                    LOG.WriteToDBUpdatesLog("--************************************")
                    LOG.WriteToDBUpdatesLog("--Update File:" + FQN)
                    LOG.WriteToDBUpdatesLog(MySql)
                    B = DB.ExecuteSql(MySql, cs, False)
                    If B Then
                        B = DB.insertDBUpdate(FQN)
                        If B Then
                            LOG.WriteToArchiveLog("DB Update applied from " + FQN)
                            B = MarkFileAsApplied(FQN, cs)
                            If B Then
                                Dim FN As String = Path.GetFileName(FQN)
                                LOG.WriteToArchiveLog("Notice DB Update:" + FN + " marked as APPLIED.")
                                DB.updateDBUpdateLastwrite(FQN)
                            End If
                        End If
                    Else
                        LOG.WriteToArchiveLog("ERROR: DB Update FAILED from " + FQN + vbCrLf + MySql)
                    End If
                    MySql = ""
                Else
                    MySql += sline + Environment.NewLine
                End If
                sline = reader.ReadLine
            Loop Until sline Is Nothing
        End Using

        If MySql.Trim.Length > 0 Then
            frmDBUpdates.txtSql.Text = Path.GetFileName(MySql)
            LOG.WriteToDBUpdatesLog("--************************************")
            LOG.WriteToDBUpdatesLog("--Update File:" + FQN)
            LOG.WriteToDBUpdatesLog(MySql)
            B = DB.ExecuteSql(MySql, cs, False)
            If B Then
                B = DB.insertDBUpdate(FQN)
                If B Then
                    LOG.WriteToArchiveLog("DB Update applied from " + FQN)
                    B = MarkFileAsApplied(FQN, cs)
                    DB.updateDBUpdateLastwrite(FQN)
                    If B Then
                        Dim FN As String = Path.GetFileName(FQN)
                        LOG.WriteToArchiveLog("Notice DB  Update: " + FN + " marked as APPLIED.")
                    End If
                End If
            Else
                LOG.WriteToArchiveLog("ERROR: DB Update FAILED from " + FQN + vbCrLf + MySql)
            End If
        End If

        B = MarkFileAsApplied(FQN, cs)

        frmDBUpdates.txtSql.Text = ""

        Return Ifailures

    End Function

    Public Function MarkFileAsApplied(FQN As String, cs As String) As Boolean
        Dim b As Boolean = True
        FQN = FQN.Replace("'", "''")
        Dim MySql As String = "update DBUpdate set UpdateApplied = 1 where FileName = '" + FQN + "'"

        b = DB.ExecuteSql(MySql, cs, False)

        Return b

    End Function


    Public Function GetFiles(ByVal Path As String, ByVal FilterList As List(Of String)) As List(Of FileInfo)
        Dim d As New DirectoryInfo(Path)
        Dim files As List(Of FileInfo) = New List(Of FileInfo)
        For Each Filter As String In FilterList
            'the files are appended to the file array
            Application.DoEvents()
            files.AddRange(d.GetFiles(Filter))
        Next
        files.Sort(f >= f.LastWritetime)
        Return files
    End Function

End Class
